sample/rules.go

```diff
@@ -149,6 +149,20 @@ func ruleMatchesTrace(t *types.Trace, rule *config.RulesBasedSamplerRule, checkN
 	var matched int
 
 	for _, condition := range rule.Conditions {
+		// This condition is evaluated for the trace as a whole.
+		// If RootSpan is nil, it means the trace timer has fired or the trace has been
+		// ejected from the cache before the root span has arrived.
+		if condition.Operator == config.HasRootSpan {
+			if (t.RootSpan != nil) == config.TryConvertToBool(condition.Value) {
```

Comment:
Is there any thread safety we need to include here?  Is there a chance the root span arrives while we're in this function?

Reply:
No; at this point the trace has been removed from the cache and we're iterating the list of traces that were removed. There is a slight gap in the time between when the traces are pulled from the cache until we make and record the trace decision, but that's a longstanding issue in refinery. It's probably the cause of the occasional missing root span where the root arrives at the exact instant it's been pulled out of cache.